src/app/fdctl/run/tiles/fd_gui.c

```diff
@@ -0,0 +1,342 @@
+}
+
+static void
+before_credit( void *             _ctx,
+               fd_mux_context_t * mux ) {
+  (void)mux;
+
+  fd_gui_ctx_t * ctx = (fd_gui_ctx_t *)_ctx;
+
+  fd_http_server_poll( ctx->gui_server );
+  fd_gui_poll( ctx->gui );
+}
+
+static inline void
+during_frag( void * _ctx,
+             ulong  in_idx,
+             ulong  seq,
+             ulong  sig,
+             ulong  chunk,
+             ulong  sz,
+             int *  opt_filter ) {
+  (void)in_idx;
+  (void)seq;
+  (void)sig;
+  (void)opt_filter;
+
+  fd_gui_ctx_t * ctx = (fd_gui_ctx_t *)_ctx;
+
+  uchar * src = (uchar *)fd_chunk_to_laddr( ctx->in_mem, chunk );
+
+   /* ... todo... sigh, sz is not correct since it's too big */
+  if( sig==FD_PLUGIN_MSG_GOSSIP_UPDATE || sig==FD_PLUGIN_MSG_VOTE_ACCOUNT_UPDATE || sig==FD_PLUGIN_MSG_VALIDATOR_INFO ) sz = 8UL + 40200UL*(58UL+12UL*34UL);
+  else if( sig==FD_PLUGIN_MSG_LEADER_SCHEDULE ) sz = 40UL + 40200UL*40UL;
+
+  if( FD_UNLIKELY( chunk<ctx->in_chunk0 || chunk>ctx->in_wmark || sz>sizeof( ctx->buf ) ) )
+    FD_LOG_ERR(( "chunk %lu %lu corrupt, not in range [%lu,%lu]", chunk, sz, ctx->in_chunk0, ctx->in_wmark ));
+
+  fd_memcpy( ctx->buf, src, sz );
+}
+
+static inline void
+after_frag( void *             _ctx,
+            ulong              in_idx,
+            ulong              seq,
+            ulong *            opt_sig,
+            ulong *            opt_chunk,
+            ulong *            opt_sz,
+            ulong *            opt_tsorig,
+            int *              opt_filter,
+            fd_mux_context_t * mux ) {
+  (void)in_idx;
+  (void)seq;
+  (void)opt_chunk;
+  (void)opt_tsorig;
+  (void)opt_filter;
+  (void)mux;
+
+  fd_gui_ctx_t * ctx = (fd_gui_ctx_t *)_ctx;
+
+  fd_gui_plugin_message( ctx->gui, *opt_sig, ctx->buf, *opt_sz );
+}
+
+static fd_http_server_response_t
+gui_http_request( fd_http_server_request_t const * request ) {
+  if( FD_UNLIKELY( request->method!=FD_HTTP_SERVER_METHOD_GET ) ) {
+    return (fd_http_server_response_t){
+      .status            = 400,
+    };
+  }
+
+  if( FD_LIKELY( !strcmp( request->path, "/websocket" ) ) ) {
+    return (fd_http_server_response_t){
+      .status            = 200,
+      .upgrade_websocket = 1,
+    };
+  } else if( FD_LIKELY( !strcmp( request->path, "/favicon.svg" ) ) ) {
+    return (fd_http_server_response_t){
+      .status            = 200,
+      .static_body       = firedancer_svg,
+      .static_body_len   = firedancer_svg_sz,
+      .content_type      = "image/svg+xml",
+      .upgrade_websocket = 0,
+    };
+  }
+
+  for( ulong i=0UL; i<sizeof(STATIC_FILES)/sizeof(STATIC_FILES[0]); i++ ) {
+    if( !strcmp( request->path, STATIC_FILES[ i ].name ) ||
+        (!strcmp( STATIC_FILES[ i ].name, "/index.html" ) &&
+          (!strcmp( request->path, "/" ) ||
+           !strcmp( request->path, "/leaderSchedule" ) ||
+           !strcmp( request->path, "/gossip" ))) ) {
```

Comment:
I think it's kind of strange to have these checks inside the loop, since they're loop invariant